rfcs/text/0017_url_service.md

```diff
@@ -0,0 +1,587 @@
+1. Expose a "redirect" endpoint, which given URL generator ID and parameters,
+   redirects the user to a deep link inside Kibana.
+1. Add ability to generate human-readable short URLs.
+1. Add ability to delete short URLs.
+1. Add ability to generate short URLs using URL generators.
+1. Will not use MD5 hashing algorithm.
+
+
+# Terminology
+
+In the proposed new service we introduce "locators". This is mostly a change
+in language, we are renaming "URL generators" to "locators".
+
+
+# Basic example
+
+The URL Service will have a client (`UrlServiceClient`) which will have the same
+interface, both, on the server-side and the client-side. It will also have a
+documented public set of HTTP API endpoints for use by: (1) the client-side
+client; (2) external users, Elastic Cloud, and Support.
+
+The following code examples will work, both, on the server-side and the
+client-side, as the base `UrlServiceClient` interface will be similar in both
+environments.
+
+Below we consider four main examples of usage of the URL Service. All four
+examples are existing use cases we currently have in Kibana.
+
+
+## Navigating within Kibana using locators
+
+In this example let's consider a case where Discover app creates a locator,
+then another plugin uses that locator to navigate to a deep link within the
+Discover app.
+
+First, the Discover plugin creates its locator (usually one per app).
+
+```ts
+const locator = plugins.share.locators.create({
+  id: 'DISCOVER_DEEP_LINKS',
+  getLocation: ({
+    indexPattern,
+    highlightedField,
+    filters: [],
+    query: {},
+    fields: [],
+    activeDoc: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx',
+  }) => {
+    app: 'discover',
+    route: `/${indexPatten}#_a=(${risonEncode({filters, query, fields})})`,
+    state: {
+      highlightedField,
+      activeDoc,
+    },
+  },
+});
+```
+
+Now, the Discover plugin exports this locator from its plugin contract.
+
+```ts
+class DiscoverPlugin() {
+  start() {
+    return {
+      locator,
+    };
+  }
+}
+```
+
+Finally, if any other app now wants to navigate to a deep link link withing the
+Discover application, they use this exported locator.
+
+```ts
+plugins.discover.locator.navigate({
+  indexPattern: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
+  highlightedField: 'foo',
+});
+```
+
+Note, in this example the `highlightedField` parameter will not appear in the
+URL bar, it will be passed to the Discover app through [`history.pushState()`](https://developer.mozilla.org/en-US/docs/Web/API/History/pushState)
+mechanism (in Kibana case, using the [`history`](https://www.npmjs.com/package/history) package, which is used by `core.application.navigateToApp`).
+
+
+## Sending a deep link to Kibana
+
+We have use cases were a deep link to some Kibana app is sent out, for example,
+through e-mail or as a Slack message.
+
+In this example, lets consider some plugin gets hold of the Discover locator
```

Comment:
i think we should rather recommend generating a short url here: `plugins.discover.locator.getShortUrl(...)` as it has the benefit of us being able to migrate those.

Reply:
I don't think I like this idea of generating a new short URL (which is a new saved object) every time somebody simply needs a deep link into some Kibana app.

1. It is slow.
2. It creates a new saved object.
3. It will not work for read-only anonymous users.
4.